### ISSUE TITLE
Need an efficient helper to check if an instance of T is a default(T)

### ISSUE THREAD

**Issue by VSadov** [dotnet/corefx#16209]
The need arises once in a while to check if a value at hand is not initialized.

See for example comments on  https://github.com/dotnet/corefx/pull/16122

Also, If/when parameterless constructors for structs are introduced, the need for `IsDefault` check could become more apparent.

**Comment by VSadov** [comment]
CC @jkotas

**Comment by jamesqo** [comment]
Yes, definitely, this would be awesome. A cost-free API for checking whether some generic type is a class (and not a nullable) would be nice too, if it can be implemented straightforwardly.

**Comment by jkotas** [comment]
> The need arises once in a while to check if a value at hand is not initialized.

Do you have more examples where this would be needed? It is not a rocket science to write such helper - take address of the value and its size, and check whether it is all zeros (using unsafe code, potentially with the usual loop unrolling, etc.).

> A cost-free API for checking whether some generic type is a class (and not a nullable) would be nice too

Agree. It should be separate issue.

**Comment by JonHanna** [comment]
If it could work on a boxed value, I'd use it in `System.Linq.Expressions.Compiler.ILGen.TryEmitConstant` so that any value type not already handled could have a constant that happened to be the default emitted the same way default expressions are. As is one would have to create a default value of the given type through reflection before using `RuntimeHelpers.Equals()` and that allocation just to test is probably not worth it.

**Comment by sakno** [comment]
I've implemented prototype of such method as a part of my library so the code can be reused. Source code is [here](https://github.com/sakno/dotNext/blob/4373fa50d0fd1ccf34d2786a88ae1c05e4523e82/src/DotNext/ValueType.cs#L268) and the public method is [here](https://github.com/sakno/dotNext/blob/4373fa50d0fd1ccf34d2786a88ae1c05e4523e82/src/DotNext/ValueType.cs#L320). 
I did that according with the following assumptions:
1. The value to be checked is passed by value in the method so I have guarantees that stack-allocated value will not be moved by GC event if value type has fields of reference types
1. `sizeof` IL instruction always can be replaced by constant value at JIT compilation. All checks in my implementation based on assumption that JIT removes dead code because all subsequent checks based on the value returned by `sizeof`.
1. The code contains a few optimized paths for data types of size 1, 2, 4 and 8. In this case JIT replaces my code with simple `ceq` operation.
1. For large structs (more than 8 bytes) I use memory aligned zero check accelerated by [Vector](https://docs.microsoft.com/en-us/dotnet/api/system.numerics.vector) if underlying hardware supports it

Additionally, I have `BitwiseEquals` method that works in similar way and it demonstrates [good performance](https://sakno.github.io/dotNext/benchmarks.html#bitwise-equality).

However, I used inline IL for implementation to keep the full control over generated code. But I'm sure that the same behavior can be achieved using `S.R.CS.Unsafe` class
